xrtpy/response/xrt_teem.py

```diff
@@ -0,0 +1,735 @@
+            print(
+                f"                    - Photon noise < "
+                f"{photon_noise_threshold*100.} %"
+            )
+    else:
+        if verbose:
+            Tmodel = tresp1.CHIANTI_temperature.value
+            print("from xrt_teem:")
+            print(f"Examined T_e range: {Tmodel.min()} - {Tmodel.max()} K")
+            print("No thresholds applied")
+    Tmap, EMmap, Terrmap, EMerrmap = make_results_maps(
+        hdr1, hdr2, T_e, EM, T_error, EMerror, mask
+    )
+    return Tmap, EMmap, Terrmap, EMerrmap
+
+
+def rebin_image(data, binfac=1):
+    """
+    Given a data array and a binning factor return the data array rebinned by
+    the binning factor. Note: the size of the original data array is
+    preserved, despite the adding up of adjacent pixels. Thus the sum of all
+    the pixels will be increased by the factor binfac.
+    """
+
+    s = data.shape
+    ns = (s[0] // binfac, s[1] // binfac)
+    rbs = (ns[0], binfac, ns[1], binfac)
+    # sums the data in binfac x binfac sized regions
+    drbin = data.reshape(rbs).sum(-1).sum(1)
+    # for a boolean mask, this makes a pixel masked if any of the summed
+    # pixels is masked. If we want to mask only if all the pixels are masked
+    # then we could use prod in place of sum above
+
+    if data.dtype == bool:
+        # need to convert back to bool after summing
+        drbin = drbin.astype(bool)
+    # This restores the image to the size of the original images as in the
+    # IDL code:
+    if data.dtype == bool:
+        dtmp = np.zeros_like(data, dtype=bool)
+    else:
+        dtmp = np.zeros_like(data)
+    for i in range(binfac):
+        for j in range(binfac):
+            dtmp[i::binfac, j::binfac] = drbin[:]
+    data = dtmp
+    return data
+
+
+def deriv(x, y):
+    """
+    Use three-point Lagrangian interpolation to compute the first derivative
+    of an array of data
+
+    Method taken from IDL routine of the same name.
+
+    Parameters:
+    -----------
+    x : float array
+        abscissa of the data
+    y : float array
+        ordinate of the data
+
+    Returns:
+    --------
+    float array :
+        first derivative at the data points
+    """
+    x0 = x[:-2]
+    x1 = x[1:-1]
+    x2 = x[2:]
+    y0 = y[:-2]
+    y1 = y[1:-1]
+    y2 = y[2:]
+    x01 = x0 - x1
+    x02 = x0 - x2
+    x12 = x1 - x2
+    dydx1 = (
+        y0 * x12 / (x01 * x02) + y1 * (1.0 / x12 - 1.0 / x01) - y2 * x01 / (x02 * x12)
+    )
+    dydx0 = (
+        y0[0] * (x01[0] + x02[0]) / (x01[0] * x02[0])
+        - y1[0] * x02[0] / (x01[0] * x12[0])
+        + y2[0] * x01[0] / (x02[0] * x12[0])
+    )
+    dydxN = (
+        -y0[-1] * x12[-1] / (x01[-1] * x02[-1])
+        + y1[-1] * x02[-1] / (x01[-1] * x12[-1])
+        - y2[-1] * (x02[-1] + x12[-1]) / (x02[-1] * x12[-1])
+    )
+    return np.append(np.insert(dydx1, 0, dydx0), dydxN)
+
+
+def derive_temperature(map1, map2, tresp1, tresp2, binfac=1, Trange=None):
```

Comment:
Hmm. At first I was thinking that it could just be available as part of the public API, but one thing would likely cause issues for users - divide by zero errors when flux ratios are calculated. Also having to instantiate the temperatureResponseFundamental objects would be a pain for people. So, yes I will rename it _derive_temperature.